### DISCUSSION TITLE
Help Testing Web Flashing

### DISCUSSION THREAD

**Discussion by witnessmenow** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#162] [Show and tell]
Howdy All,

I've been playing around with the [ESP Web Tools](https://esphome.github.io/esp-web-tools/) and I think it could be really useful for sharing of projects in the future. Basically it allows your to flash a binary directly to your ESP without needing to have installed any of the libraries that make up the sketch.

If anyone would be interested in testing something for me it would be really helpful. I've set up a test page for my WiFi tetris clock project, where it displays the time using falling tetris blocks

Demo clip of it here:
https://content.instructables.com/ORIG/FS0/N0XF/KP41SN17/FS0N0XFKP41SN17.gif?format=mp4

**You would need** 
- An ESP32 connected to a 64x32 matrix panel using the default pins of the library (or a 64x64 where e is connected to 18)
- A computer with chrome

https://witnessmenow.github.io/ESP32-Trintiy-Web-Flash/WifiTetrisClock/

If you visit this page in Chrome you should be able to flash the WiFi Tetris clock sketch directly onto your ESP32 once it plugged in.

The sketch has WiFi manager included in it, so it will launch a config portal where you can config the clock and matrix display.

There are some clock configuration options in the config portal as well as some Matrix ones. If people wanted to try it out and let me know what you think I would really appreciate it.

(I'll probably upload a simpler example later, but I wanted to try a relatively complicated one first to prove the solution could cover all bases)

**Comment by donnersm** [comment]
Hi Brian,
That looks promising... I tested it with my esp32 hardware that i used in my FFT analyser. Here is what i noticed:
- Upload from website seems to work fine. ESP is flashing and by using the serial monitor i see the output of your sketch
- After connecting to the Tetriswife, I see the config page although i had some trouble finding it.. maybe it is an idea to show the ip adres somewhere, like in the browser... I know its in the serial monitor but. he... that wasn't the first place i looked
- The display is doing somthing i see some timed movement however what i see is not correct. I'm sure the hardware pinout between your and mine design are different. that also what i miss:
- We should be able to change the pinout in the configuration page.
Also once I configured the local wifi and save the config file, reboot etc...the clock was connected to my network but I could;t find any way to connect to the config panel again...
But again..this looks promising.
Hope this helps

**Reply by witnessmenow** [reply]
Hey @donnersm , thanks for the reply

- I've updated the code to display the SSID and IP on the display (if it can). 
Were you using a PC? I've been testing on my phone and it does the airport/hotel wifi thing where as soon as I connect to the conf SSID it brings me to the "sign in" page. I will improve the webflash instructions in the future so could add the IP there too.

- Are you using the default pins of this library for your display board? It would be possible to add the ability to update all the pins, but it would be a lot of things to add to config portal! The main purpose of the web flashing thing for me is to support some hardware that I make (A shield for the D1 mini style ESP32 boards and soon to be released All in one board) so this wouldn't be a huge priority for me. If you are interested in the code for your design, you can find the code for the clock here: https://github.com/witnessmenow/ESP32-Trinity/blob/master/examples/WifiTetrisClock/WifiTetrisClock.ino(this is the version without the config portal, will add the config portal one to the same repo later tonight probably) The config portal code makes the sketch a lot messier!
- Something else for the documentation! If you press the reset button of your ESP32 twice about a second apart it will force config mode. The other reasons it will enter config mode:
    - Can not connect to your last known WiFi
    - There is no config on the spiffs for the project 
    
Thanks for the feedback!

If it helps at all, I wrote up instructions on how to create the webflash thingy up, its actually pretty straight forward. Will probably make a video on it soon. I saw your FFT analyzer project here a few weeks ago, it looks amazing! Is the code released for it yet?

EDIT: forgot the link to webflash write up! https://github.com/witnessmenow/ESP-Web-Tools-Tutorial

**Comment by donnersm** [comment]
Yes , the fft code is here:
https://github.com/donnersm/FFT_ESP32_Analyzer
